text/0005-syncing.md

```diff
@@ -0,0 +1,114 @@
+3. Synchronize all missing dependencies.
+4. Execute deploys from the linear chain in the order of them appearing in the linear chain.
+
+`BlockExecutor` sees whether it's a `switch_block` and whether it should run an auction for era `era_id + AUCTION_DELAY` . As a result of execution, it will produce a `Block` instance (that is a block of the linear chain) but also an event `NextEraValidators(finalized_block.era_id + 1, Set<Validator>)` that informs Consensus about the next era validators. Consensus component will use new validators (and random bits from blocks between booking block and key block of the next era) and create an instance of new era.
+
+Same mechanism can be used in a node that is catching up – it will use `Block`s from the linear chain to build up its Global State via `BlockExecutor` (we need a mechanism of turning `Block` into `FinalizedBlock` as that's the input that `BlockExecutor` understands) and create new eras. 
+
+Obviously, over the course of synchronization some eras will become obsolete but the assumption is that eras are deactivated/removed by a separate mechanism. For example, `BlockExecutor` component already knows about the unbonding delay so it could send an event to a consensus component that it should deactivate an era (or if consensus component needs this knowledge it could deactivate old eras by itself).
+
+#### Variant B ####
+
+Since downloading the whole linear chain (and most importantly, the deploys) is a time-consuming operation, a faster alternative would be to just sync the Global State at the initial hash. Global State is a merkle trie so it's easily splittable and has an easy way of verifying its correctness (compare root hashes).
+
+If we choose to synchronize Global State directly though, the above won't work. We need to be able to create active eras just by looking at the Global State. This should also be fairly trivial since PoS contract can track past eras, current ones and future (auctions already executed but era not yet started). We can use information from the cache and create instances of eras that are still active.
+
+**Necessary work:**
+
+1. Start a node with the trusted hash
+2. Ask peers for the initial block from the linear chain.
+3. Synchronize Global State at the post-state hash of the initial block.
+4. Validate correctness of the Global State with the post-state hash from the initial block.
+
+
+---------------------
+Once whole linear chain is "consumed" and state is built-up, we need to synchronize DAG(s) of still active eras.
+
+### Synchronizing DAG(s) ###
+
+#### Variant A (quick-n-dirty) ####
+
+Once node has built up its Global State and has necessary eras started it could rely on already existing mechanism of syncing nodes – i.e. it could transition into an *active* node (not a validator) and wait for new messages. Upon receiving a message for which it doesn't have all the necessary dependencies it will already start synchronizing.
```

Comment:
It sounds like that could delay passing messages between honest validators by a lot, because one of the intermediate hops has to download the whole DAG to verify that one message. I guess in the long run, only variant B is viable?

Reply:
Yes, it's _quick-n-dirty_ variant that requires least effort to add the sync feature.

Reply:
I hope it won't be _that_ slow. The main contributor will be deploys that have to be downloaded…

Reply:
For our projected gamma (charlie?) net, I would assume this is not an issue due to the small size? With low traffic, the Global State is small, and with shorter eras there's an upper bound on DAG size, I would assume.

Reply:
Messages are gossiped, right? So even if someone starts downloading the whole DAG, others would still be gossiping the message and it could reach everyone else before this someone finishes... no?

Reply:
I think that Andreas' meant that if one of the "infection targets" is a syncing node then the ability to gossip new vertices to active validators is limited since part of the traffic is going through nodes that won't gossip (syncing node cannot gossip since we must gossip only valid messages – i.e. ones that we can validate).

Reply:
Okay, it will be slightly limited for a moment, but we never infect just one node, right? And the rest will keep infecting others, so in general the setback shouldn't be too big.

At least from one node. If we have multiple syncing nodes, it might get worse.